aten/src/ATen/native/cuda/fused_adam_utils.cuh

```diff
@@ -102,90 +102,93 @@ C10_DEVICE __forceinline__ void adam_math(
 // parameter updates accordingly. To be functionally on par with `torch.optim`
 // optimizers and `_multi_tensor` ones, the kernel below writes out gradients
 // only when `grad_scale_ptr != nullptr.
-template <typename scalar_type, int depth = 4>
+template <typename scalar_type, int depth, ADAM_MODE adam_mode, bool amsgrad>
 struct FusedAdamMathFunctor {
   static_assert(
       depth == 4 || depth == 5,
       "depth of 4 for Adam, depth of 5 for Adam with AMSGrad.");
   using opmath_t = at::opmath_type<scalar_type>;
-  C10_DEVICE __forceinline__ void operator()(
-      int chunk_size,
+  template <typename index_t>
+  C10_DEVICE inline void operator()(
+      const index_t& chunk_size,
       FusedOptimizerTensorListMetadata<depth>& tl,
       const float* lr_ptr,
-      const double lr,
-      const double beta1,
-      const double beta2,
-      const double weight_decay,
-      const double eps,
-      const bool maximize,
-      const bool amsgrad,
+      const double& lr,
+      const double& beta1,
+      const double& beta2,
+      const double& weight_decay,
+      const double& eps,
+      const bool& maximize,
       const float* grad_scale_ptr,
-      const float* found_inf_ptr,
-      const ADAM_MODE adam_mode) {
-    int tensor_loc = tl.block_to_tensor[blockIdx.x];
-    int chunk_idx = tl.block_to_chunk[blockIdx.x];
-    int n = tl.numel_for_tensor[tensor_loc];
-    double lr_double = lr_ptr ? *lr_ptr : lr;
+      const float* found_inf_ptr) {
+    const index_t tensor_loc = tl.block_to_tensor[blockIdx.x];
+    const index_t chunk_idx = tl.block_to_chunk[blockIdx.x];
+    const double lr_double = lr_ptr ? *lr_ptr : lr;
 
     if (found_inf_ptr && *found_inf_ptr == 1) {
       return;
     }
-    auto* step_count =
-        reinterpret_cast<const float*>(tl.state_steps_addresses[tensor_loc]);
+    const auto [bias_correction1, bias_correction2_sqrt] =
```

Comment:
Another q for my understanding: I observe two differences between this new impl and the old. Firstly, we've now moved the bias correction calculation after we've `reinterpret_cast` step_count. Secondly, we compute them in a lambda now versus just inlining them as before. I can kinda see how the first could help (though if you could explain that would help me more) but I have no inkling why the second would help.